api/src/main/java/org/openmrs/Cohort.java

```diff
@@ -302,29 +361,22 @@ public void setName(String name) {
 	}
 
 	public Set<Integer> getMemberIds() {
-		memberIds = new TreeSet<Integer>();
-		for (CohortMembership member : getMembers()) {
-			memberIds.add(member.getPatient().getPatientId());
+		Set<Integer> memberIds = new TreeSet<Integer>();
+		for (CohortMembership member : getMemberships()) {
+			memberIds.add(member.getPatientId());
 		}
 		return memberIds;
 	}
 
 	public void setMemberIds(Set<Integer> memberIds) {
-		this.memberIds = new TreeSet<Integer>(memberIds);
-		for (Integer id : memberIds) {
-			addMembership(new CohortMembership(new Patient(id)));
-		}
-	}
-
-	public Collection<CohortMembership> getMembers() {
-		if (members == null) {
-			members = new TreeSet<CohortMembership>();
+		Set<Integer> ids = new TreeSet<Integer>(memberIds);
```

Comment:
Isn't this going to cause issues with hibernate if this is in the session? Because hibernate will have a different instance of the collection from the new one you are creating here. You should have this method just assign the new value. Yours and client code should never call this but instead addMember()